mypy/checkexpr.py

```diff
@@ -3518,10 +3519,71 @@ def visit_dict_expr(self, e: DictExpr) -> Type:
                         variables=[kt, vt])
                     rv = self.check_call(constructor, [arg], [nodes.ARG_POS], arg)[0]
                 else:
-                    self.check_method_call_by_name('update', rv, [arg], [nodes.ARG_POS], arg)
+                    arg_type = arg.accept(self)
+                    rv = self.dict_unpack(rv, arg_type, arg)
         assert rv is not None
         return rv
 
+    def dict_unpack(self, dict_obj: Type, other: Type, context: Context) -> ProperType:
+        """Joins two dict-like objects.
+
+        For example, `dict[str, int]` and `dict[int, int]`
+        will become `dict[object, int]`.
+        """
+        dict_obj = get_proper_type(dict_obj)
+        if (not isinstance(dict_obj, Instance)
+                or dict_obj.type.fullname != 'builtins.dict'
+                or len(dict_obj.args) != 2):
+            # Since this function is only used when two dicts are joined like:
+            # `{**other, 'a': 1}`, we require `dict_obj` to be an `Instance`
+            # of `builtins.dict``. This should not happen, unless someone else
+            # will call this function from other places.
+            return AnyType(TypeOfAny.from_error)
```

Comment:
How about using an assert here (well, only if it's not an Instance) instead, so that misuse of the function gets caught? (Though there's an argument to be made that anything calling this function might not be tested enough and asserts at runtime kinda suck)

Reply:
Yes, I will go with `assert` instead 👍